provisioner/node_pools.go

```diff
@@ -0,0 +1,357 @@
+package provisioner
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha512"
+	"encoding/base64"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+	"text/template"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/cloudformation"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando-incubator/cluster-lifecycle-manager/api"
+	awsExt "github.com/zalando-incubator/cluster-lifecycle-manager/pkg/aws"
+	"github.com/zalando-incubator/cluster-lifecycle-manager/pkg/updatestrategy"
+)
+
+const (
+	userDataFileName      = "userdata.clc.yaml"
+	stackFileName         = "stack.yaml"
+	nodePoolTagKeyLegacy  = "NodePool"
+	nodePoolTagKey        = "kubernetes.io/node-pool"
+	nodePoolRoleTagKey    = "kubernetes.io/role/node-pool"
+	nodePoolProfileTagKey = "kubernetes.io/node-pool/profile"
+)
+
+// NodePoolProvisioner is able to provision node pools for a cluster.
+type NodePoolProvisioner interface {
+	Provision(values map[string]string) error
+	Reconcile() error
+}
+
+// AWSNodePoolProvisioner is a node provisioner able to provision node pools
+// in AWS via cloudformation.
+type AWSNodePoolProvisioner struct {
```

Comment:
Based on the request for baremetal support, we should have a TODO here to refactor these types and methods into providers.

Reply:
Added a TODO.